tracing/opencensus/http.go

```diff
@@ -0,0 +1,174 @@
+package opencensus
+
+import (
+	"context"
+	"net/http"
+
+	"go.opencensus.io/plugin/ochttp"
+	"go.opencensus.io/plugin/ochttp/propagation/b3"
+	"go.opencensus.io/trace"
+
+	kithttp "github.com/go-kit/kit/transport/http"
+)
+
+// HTTPClientTrace enables OpenCensus tracing of a Go kit HTTP transport client.
+func HTTPClientTrace(options ...TracerOption) kithttp.ClientOption {
+	cfg := TracerOptions{
+		sampler:       trace.AlwaysSample(),
+		httpPropagate: &b3.HTTPFormat{},
+	}
+
+	for _, option := range options {
+		option(&cfg)
+	}
+
+	clientBefore := kithttp.ClientBefore(
+		func(ctx context.Context, req *http.Request) context.Context {
+			var name string
+
+			if cfg.name != "" {
+				name = cfg.name
+			} else {
+				// OpenCensus states Path being default naming for a client span
+				name = req.Method + " " + req.URL.Path
+			}
+
+			span := trace.NewSpan(
+				name,
+				trace.FromContext(ctx),
+				trace.StartOptions{
+					Sampler:  cfg.sampler,
+					SpanKind: trace.SpanKindClient,
+				},
+			)
+
+			span.AddAttributes(
+				trace.StringAttribute(ochttp.HostAttribute, req.URL.Host),
+				trace.StringAttribute(ochttp.MethodAttribute, req.Method),
+				trace.StringAttribute(ochttp.PathAttribute, req.URL.Path),
+				trace.StringAttribute(ochttp.UserAgentAttribute, req.UserAgent()),
+			)
+
+			if !cfg.public {
+				cfg.httpPropagate.SpanContextToRequest(span.SpanContext(), req)
+			}
+
+			return trace.NewContext(ctx, span)
+		},
+	)
+
+	clientAfter := kithttp.ClientAfter(
+		func(ctx context.Context, res *http.Response) context.Context {
+			if span := trace.FromContext(ctx); span != nil {
+				span.SetStatus(ochttp.TraceStatus(res.StatusCode, http.StatusText(res.StatusCode)))
+				span.AddAttributes(
+					trace.Int64Attribute(ochttp.StatusCodeAttribute, int64(res.StatusCode)),
+				)
+			}
+			return ctx
+		},
+	)
+
+	clientFinalizer := kithttp.ClientFinalizer(
+		func(ctx context.Context, err error) {
+			if span := trace.FromContext(ctx); span != nil {
+				if err != nil {
+					span.SetStatus(trace.Status{
+						Code:    trace.StatusCodeUnknown,
+						Message: err.Error(),
+					})
+				}
+				span.End()
+			}
+		},
+	)
+
+	return func(c *kithttp.Client) {
+		clientBefore(c)
+		clientAfter(c)
+		clientFinalizer(c)
+	}
+}
+
+// HTTPServerTrace enables OpenCensus tracing of a Go kit HTTP transport server.
+func HTTPServerTrace(options ...TracerOption) kithttp.ServerOption {
```

Comment:
Really cool use of transport/http.ServerOption, lots of stuff packed in to these functions.